Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m

```diff
@@ -85,30 +106,55 @@ + (NSString *)cellReuseIdentifier
 
 - (void)loadForDisplayWithTransaction:(YapDatabaseReadTransaction *)transaction
 {
+    OWSAssert(self.layoutInfo);
+    OWSAssert(self.layoutInfo.viewWidth > 0);
     OWSAssert(self.viewItem);
     OWSAssert([self.viewItem.interaction isKindOfClass:[OWSContactOffersInteraction class]]);
 
+    [self configureFonts];
+
     OWSContactOffersInteraction *interaction = (OWSContactOffersInteraction *)self.viewItem.interaction;
 
     OWSAssert(
         interaction.hasBlockOffer || interaction.hasAddToContactsOffer || interaction.hasAddToProfileWhitelistOffer);
 
-    [self setNeedsLayout];
-}
-
-- (UIFont *)titleFont
-{
-    return UIFont.ows_dynamicTypeBodyFont.ows_mediumWeight;
-}
-
-- (UIFont *)buttonFont
-{
-    return UIFont.ows_dynamicTypeBodyFont;
-}
+    CGFloat buttonWidth = 0.f;
+    if (interaction.hasAddToContactsOffer) {
+        [self.stackView addArrangedSubview:self.addToContactsButton];
+        buttonWidth = MAX(buttonWidth, [self.addToContactsButton sizeThatFits:CGSizeZero].width);
+    }
+    if (interaction.hasAddToProfileWhitelistOffer) {
+        [self.stackView addArrangedSubview:self.addToProfileWhitelistButton];
+        buttonWidth = MAX(buttonWidth, [self.addToProfileWhitelistButton sizeThatFits:CGSizeZero].width);
+    }
+    if (interaction.hasBlockOffer) {
+        [self.stackView addArrangedSubview:self.blockButton];
+        buttonWidth = MAX(buttonWidth, [self.blockButton sizeThatFits:CGSizeZero].width);
+    }
 
-- (CGFloat)hMargin
-{
-    return 10.f;
+    buttonWidth = (2 * self.buttonHPadding + (CGFloat)ceil(buttonWidth));
+
+    CGFloat hMargins = (self.layoutInfo.fullWidthGutterLeading + self.layoutInfo.fullWidthGutterTrailing);
+    CGFloat maxButtonWidth = self.layoutInfo.viewWidth - hMargins;
+    buttonWidth = MIN(buttonWidth, maxButtonWidth);
+
+    CGFloat buttonHeight = self.buttonHeight;
+    [NSLayoutConstraint deactivateConstraints:self.layoutConstraints];
+    self.layoutConstraints = @[
+        [self.addToContactsButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+        [self.addToProfileWhitelistButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+        [self.blockButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+
+        [self.addToContactsButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
+        [self.addToProfileWhitelistButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
+        [self.blockButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
```

Comment:
I might rework this to use a second stack view, which would hopefully let us skip this sizing.